app/src/main/java/crux/bphc/cms/activities/TokenActivity.kt

```diff
@@ -45,232 +28,74 @@ class TokenActivity : AppCompatActivity() {
-        val call = moodleServices.fetchUserDetail(token)
-        call.enqueue(object : Callback<UserDetail> {
-            override fun onResponse(call: Call<UserDetail>, response: Response<UserDetail>) {
-                if (response.isSuccessful) {
-                    var userDetail: UserDetail
-                    if (response.body().also { userDetail = it!! } != null) {
-                        userDetail.token = token
-                        userDetail.privateToken = privateToken
-
-                        UserAccount.setUser(userDetail)
-                        fetchUserData()
-                    }
-                }
-            }
+        binding = ActivityTokenBinding.inflate(layoutInflater)
+        setContentView(binding.root)
 
-            override fun onFailure(call: Call<UserDetail>, t: Throwable) {
-                Log.wtf(TAG, t)
-                showProgress(false, "")
-            }
-        })
-    }
-
-    private fun onLogin() {
-        /*
-            We'll just create an into a specific Moodle endpoint. The Moodle website will handle the authentication,
-            generate a token and redirect the browser to the Uri with specified schema. The browser will create an
-            intent that'll launch this activity again.
-         */
-
-        // A random number that identifies the request
-        val passport = Random().nextInt(1000).toString()
-        var builder = Urls.SSO_LOGIN_URL.buildUpon()
-        builder = builder.appendOrSetQueryParameter("passport", passport)
-        builder = builder.appendOrSetQueryParameter("urlscheme", Urls.SSO_URL_SCHEME)
-        val loginUrl = builder.toString()
+        binding.submitButton.setOnClickListener { viewModel.onLogin() }
 
-        // Set the launch data, we need this to verify the token obtained after SSO
-        val data = HashMap<String, String>()
-        // SITE_URL must not end with trailing /
-        data["SITE_URL"] = Urls.MOODLE_URL.toString().replace("/$".toRegex(), "")
-        data["PASSPORT"] = passport
-        MyApplication.instance.setLoginLaunchData(data)
-        val intent = Intent(Intent.ACTION_VIEW)
-        intent.data = Uri.parse(loginUrl)
-        startActivity(intent)
+        viewModel.onErrorMessage.observe(this, onErrorMessageObserver)
+        viewModel.showBadTokenDialog.observe(this, showBadTokenDialogObserver)
+        viewModel.status.observe(this, statusObserver)
+        viewModel.startMainActivity.observe(this, startMainActivityObserver)
+        viewModel.startIntent.observe(this, startIntentObserver)
     }
 
-    private fun showProgress(show: Boolean, message: String) {
-        if (show) progressDialog.show() else progressDialog.hide()
-        progressDialog.setMessage(message)
-    }
-
-    private fun dismissProgress() {
-        if (progressDialog.isShowing) {
-            progressDialog.dismiss()
+    private val onErrorMessageObserver = Observer<SingleLiveEvent<String>> { error ->
```

Comment:
I prefer putting instance var and vals above instance methods. Not to mention, these are being referenced in `onCreate`. Lmk what you feel about making this change.

Reply:
For Observers, it's convention to keep them below onCreate since they manipulate the UI, however, all other instance vals and vars should be kept above the instance methods.

Reply:
Alright, as long it's an established convention.